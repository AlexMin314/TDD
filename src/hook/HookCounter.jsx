import React, { useState, useCallback } from "react";
import { useFetchData, testApiCall } from "./useFetchData";

export const url = "https://randomuser.me/api/";

const HookCounter = () => {
  const [number, setNumber] = useState(0);
  const [fetchName, setFetchName] = useState("Fetch");
  const [data, setData] = useState(undefined);

  const onIncrease = useCallback(() => {
    setNumber(number + 1);
  }, [number]);
  const onDecrease = useCallback(() => {
    setNumber(number - 1);
  }, [number]);

  const onFetch = () => {
    setData("Fetching...");
    setFetchName("Wait");
    testApiCall(url, setData, () => setFetchName("Fetch"));
  };

  return (
    <div>
      <h2>{number}</h2>
      <button onClick={onIncrease}>+1</button>
      <button onClick={onDecrease}>-1</button>
      <div>
        <b>Fetch on click</b>
      </div>
      <button onClick={onFetch}>{fetchName}</button>
      <div>{JSON.stringify(data)}</div>
      <div>
        <b>Fetch on mount</b>
      </div>
      <div>{JSON.stringify(useFetchData(url).data)}</div>
    </div>
  );
};

export default HookCounter;
