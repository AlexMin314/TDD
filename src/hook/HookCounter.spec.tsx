import React from 'react';
import { mount } from 'enzyme';
import HookCounter, { url } from './HookCounter';
import { render, fireEvent, waitForElement, act } from '@testing-library/react';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

const mock = new MockAdapter(axios);

jest.useFakeTimers();

describe('<HookCounter />', () => {
  it('matches snapshot', () => {
    const wrapper = mount(<HookCounter />);
    expect(wrapper).toMatchSnapshot();
  });
  it('increases', () => {
    const wrapper = mount(<HookCounter />);
    let plusButton = wrapper.findWhere(
      node => node.type() === 'button' && node.text() === '+1'
    );
    plusButton.simulate('click');
    plusButton.simulate('click');

    const number = wrapper.find('h2');

    expect(number.text()).toBe('2');
  });
  it('decreases', () => {
    const wrapper = mount(<HookCounter />);
    let decreaseButton = wrapper.findWhere(
      node => node.type() === 'button' && node.text() === '-1'
    );
    decreaseButton.simulate('click');
    decreaseButton.simulate('click');

    const number = wrapper.find('h2');

    expect(number.text()).toBe('-2');
  });
});

describe('<HookCounter />', () => {
  afterEach(() => {
    mock.reset();
  });

  it('matches snapshot', () => {
    const wrapper = mount(<HookCounter />);
    expect(wrapper).toMatchSnapshot();
  });
  it('increases', () => {
    const { getByText } = render(<HookCounter />);
    const displayNumber = getByText('0');
    const increaseButton = getByText('+1');
    fireEvent.click(increaseButton);
    fireEvent.click(increaseButton);
    expect(displayNumber).toHaveTextContent('2');
  });
  it('decreases', () => {
    const { getByText } = render(<HookCounter />);
    const displayNumber = getByText('0');
    const decreasesButton = getByText('-1');
    fireEvent.click(decreasesButton);
    fireEvent.click(decreasesButton);
    expect(displayNumber).toHaveTextContent('-2');
  });

  test('async', async () => {
    const { getByText, getByTestId } = render(<HookCounter />);

    const fetchButton = getByTestId('fetchButton');

    mock.onGet(url).reply(200, { data: { results: [{ name: 'test' }] } });

    fireEvent.click(fetchButton);

    // act(() => {
    //   jest.runAllTimers();
    // });

    await waitForElement(() => getByText('Fetch!'));
    expect(mock.history.get.length).toBe(2);
  });
});
