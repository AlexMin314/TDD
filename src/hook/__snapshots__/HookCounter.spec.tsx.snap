// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HookCounter /> matches snapshot 1`] = `
<HookCounter>
  <div>
    <h2>
      0
    </h2>
    <button
      onClick={[Function]}
    >
      +1
    </button>
    <button
      onClick={[Function]}
    >
      -1
    </button>
    <div>
      <b>
        Fetch on click
      </b>
    </div>
    <button
      onClick={[Function]}
    >
      Fetch
    </button>
    <div />
    <div>
      <b>
        Fetch on mount
      </b>
    </div>
    <div />
  </div>
</HookCounter>
`;

exports[`<HookCounter /> matches snapshot 2`] = `
<HookCounter>
  <div>
    <h2>
      0
    </h2>
    <button
      onClick={[Function]}
    >
      +1
    </button>
    <button
      onClick={[Function]}
    >
      -1
    </button>
    <div>
      <b>
        Fetch on click
      </b>
    </div>
    <button
      onClick={[Function]}
    >
      Fetch
    </button>
    <div />
    <div>
      <b>
        Fetch on mount
      </b>
    </div>
    <div />
  </div>
</HookCounter>
`;
